### PR TITLE
Catalaog upload feature

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Main catalog item upload functionality. Related issue: (https://github.com/vmware/go-vcloud-director/issues/50)

- Added upload functionality for catalog item. Now it supports ova standard files and chunked ones. Also returns task for vCD ova import.
Missing functionality which will be implemented later:
* Error handling return within xml of response
* Utilise existing Task mechanism.
* Report percentage progress
* Remove lock to xml first element
* Better error handling
* Missing tests
* Verify that "sanitizedName" func is enough to cover security issue -> See vmware/pyvcloud#268.

### CODE REVIEW

File: govcd/catalog.go [hunk 1]
```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
...
+		return Envelope{}, err
+	}
+
+	var buf bytes.Buffer
+	ovfReader := io.TeeReader(openedFile, &buf)
+
+	request := client.NewRequest(map[string]string{}, "PUT", *ovfUploadUrl, ovfReader)
+	request.Header.Add("Content-Type", "text/xml")
+
+	response, err := client.Http.Do(request)
```
Comment (dataclouder) [inline govcd/catalog.go]: The rest of the library uses a function from api.go
```response, err := checkResp(client.Http.Do(request))```

And likewise, the response is unmarshaled using ```err = decodeBody(response, destinationObject)```
This feature should use the same approach, as it allows us to handle errors from a central place.

File: govcd/tar.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
+package govcd
```
Comment (dataclouder) [inline govcd/tar.go]: We should create a separate package for functionalities that are not part of the library (file handling, compression/decompression utilities, logging, etc.)
I suggest calling it "util"

File: govcd/catalog.go [hunk 3]
```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
...
+		tempVar, err := uploadFile(client, uploadHREF, filePath, uploadedBytes, totalBytesToUpload)
+		if err != nil {
+			return err
+		}
+		uploadedBytes += tempVar
+	}
+	return nil
+}
+
+/** Function waits until vCD provides temporary file upload links. */
```
Comment (hodgesrm) [inline govcd/catalog.go]: Please use // style comments.

File: govcd/catalog.go [hunk 4]
```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
 
 	return CatalogItem{}, fmt.Errorf("can't find catalog item: %s", catalogitem)
 }
+
+/**
```
Comment (hodgesrm) [inline govcd/catalog.go]: Please convert to // style of comments to match existing code.  Check Go system packages for examples--we want to match their style both in terms of syntax as well as the content of documentation. 

More generally the comment here should describe the user-visible interface specification.  In other words--what does this function accomplish for the client, what outputs/errors does it generate?  Please push implementation details (which are very helpful) into the implementation code or routines it calls.

File: govcd/catalog.go [hunk 5]
```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
...
+		}
+	}
+	return ""
+}
+
+/** Initiates creation of item and returns ovf upload url for created item. */
+func createItemForUpload(client *Client, createHREF *url.URL, catalogItemName string, itemDescription string) (*url.URL, error) {
+
+	reqBody := bytes.NewBufferString(
+		"<UploadVAppTemplateParams xmlns=\"http://www.vmware.com/vcloud/v1.5\" name=\"" + catalogItemName + "\" >" +
```
Comment (hodgesrm) [inline govcd/catalog.go]: Can you check the docs to ensure this xmlns tag is still valid? v1.5 may cause problems in Vulcan.  We have dropped support for server API versions below 19.0.

File: govcd/catalog.go [hunk 6]
```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
...
+
+	ovfUploadUrl, err := url.ParseRequestURI(catalogItemParsed.Entity.HREF)
+	if err != nil {
+		return nil, err
+	}
+
+	return ovfUploadUrl, nil
+}
+
+/** Create Request with right headers and range settings. Support multi part file upload. */
```
Comment (hodgesrm) [inline govcd/catalog.go]: Comment to // style.

File: govcd/catalog.go [hunk 7]
```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
...
+// TODO Creates a new file upload http request with optional extra params
```
Comment (hodgesrm) [inline govcd/catalog.go]: Is this really a TODO?

File: govcd/tar.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	"archive/tar"
+	"errors"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func Unpack(tarFile string) ([]string, error) {
```
Comment (hodgesrm) [inline govcd/tar.go]: Is this meant to be part of the public API?  If not the name should be lower case.

File: govcd/tar.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+	var filePaths []string
+
+	reader, err := os.Open(tarFile)
+	if err != nil {
+		return filePaths, err
+	}
+	defer reader.Close()
+
+	tarReader := tar.NewReader(reader)
```
Comment (hodgesrm) [inline govcd/tar.go]: Does this code block out-of-tree references? That's a vuln we had to fix in pyvcloud.  See https://github.com/vmware/pyvcloud/pull/268.

File: govcd/catalog.go (hunk 3 above)
Comment (vbauzysvmware) [reply govcd/catalog.go]: Fixed

File: govcd/catalog.go (hunk 4 above)
Comment (vbauzysvmware) [reply govcd/catalog.go]: Fixed

File: govcd/catalog.go (hunk 6 above)
Comment (vbauzysvmware) [reply govcd/catalog.go]: Fixed

File: govcd/catalog.go (hunk 1 above)
Comment (vbauzysvmware) [reply govcd/catalog.go]: Fixed

File: govcd/tar.go (hunk 2 above)
Comment (vbauzysvmware) [reply govcd/tar.go]: As we got confirmation in email -> made the change.

File: govcd/catalog.go (hunk 7 above)
Comment (vbauzysvmware) [reply govcd/catalog.go]: Removed comment of todo. Added to my notes.

File: govcd/catalog.go [hunk 10]
```diff
@@ -23,6 +35,15 @@ func NewCatalog(c *Client) *Catalog {
 	}
 }
 
+type Envelope struct {
+	File []struct {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/catalog.go]: I am not sure about it, as it's file xml not API responses.

File: govcd/catalog.go [hunk 11]
```diff
@@ -55,3 +66,243 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
...
+	}
+
+	vAppTemplateParsed := &types.VAppTemplate{}
+
+	err = xml.Unmarshal(body, &vAppTemplateParsed)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("[TRACE] Response: %#v\n", response)
```
Comment (dataclouder) [inline govcd/catalog.go]: Please note that PR #76 will log HTTP requests and responses automatically, if you use `NewRequest` and `decodeBody` from api.go.

File: govcd/catalog.go (hunk 5 above)
Comment (vbauzysvmware) [reply govcd/catalog.go]: According vCloud API 30.0 seems fine: https://code.vmware.com/apis/287/vcloud#/doc/doc/types/UploadVAppTemplateParamsType.html

File: govcd/tar.go (hunk 8 above)
Comment (vbauzysvmware) [reply govcd/tar.go]: Well it are different code files "catalog" and "util/tar"

File: govcd/catalog.go (hunk 11 above)
Comment (vbauzysvmware) [reply govcd/catalog.go]: Nice. Now it's racing ;-D

File: govcd/tar.go (hunk 9 above)
Comment (vbauzysvmware) [reply govcd/tar.go]: Yes, we have similar additional functionality:
func sanitizedName(filename string) string {
	if len(filename) > 1 && filename[1] == ':' {
		filename = filename[2:]
	}
	filename = strings.Replace(filename, "\\/.", "\\/", -1)
	filename = strings.Replace(filename, "../", "", -1)
	return strings.Replace(filename, "..\\", "", -1)
}
Though I will add verify task of this security issue

**Comment by lvirbalas** [comment]
Thanks for the improvements! Given that you're going out to vacation, please merge the code, only keep the issue #50 open. We'll be looking forward to your final PR implementing the missing pieces after you're back.

**Comment by vbauzysvmware** [comment]
We merge this PR due reason I go to vacations. After return new PR will come with left functionality and tests.

**Comment by vbauzysvmware** [comment]
We merge this PR due reason I go to vacations. After return new PR will come with left functionality and tests.

**Pull request merged**